cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -158,15 +184,29 @@ public void deleteApplications(String namespaceId) {
   public void updateAppSpec(String namespaceId, String appId, String versionId, ApplicationSpecification spec) {
     LOG.trace("App spec to be updated: id: {}: spec: {}", appId, GSON.toJson(spec));
     MDSKey key = new MDSKey.Builder().add(TYPE_APP_META, namespaceId, appId, versionId).build();
+    MDSKey versionLessKey = null;
     ApplicationMeta existing = getFirst(key, ApplicationMeta.class);
+    ApplicationMeta updated;
+
+    // Check again without the version to account for old data format if might not have been upgraded yet
+    if (existing == null && (versionId.equals(ApplicationId.DEFAULT_VERSION))) {
```

Comment:
what if `existing != null && (versionId.equals(ApplicationId.DEFAULT_VERSION)`, `versionLessKey` will stay null and won't be deleted later